### PR TITLE
Add logging module to load log configuration dynamically

### LINKED ISSUE #7114: Add logging level switching API in OAP backend for debugging purpose
For now, we have some logs that are very helpful in debugging and troubleshooting but too verbose to be printed in production environment, hence we print them in `DEBUG` level.

However, switching the log level requires OAP to restart and that may make the errors recover from a restart, so in this issue, we're going to discuss and implement a new API that allows users to switch a particular logger's level. This is kind of similar to [Spring Boot Actuator loggers endpoint](https://docs.spring.io/spring-boot/docs/current/reference/html/actuator.html#actuator.loggers).

Another doc provided by @hanahmily is [log4j auto-config](https://logging.apache.org/log4j/2.x/manual/customconfig.html)

### PR DESCRIPTION
**Pull request by hanahmily**
### Logging configuration dynamic loading
- [x] Update the documentation to include this new feature.
- [x] Local testing.

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #7114 .
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

### CODE REVIEW

File: oap-server/server-logging/pom.xml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        <dependency>
+            <groupId>org.apache.skywalking</groupId>
+            <artifactId>server-core</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-api</artifactId>
+        </dependency>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (hanahmily) [reply oap-server/server-logging/pom.xml:42]: ignore

**Comment by wu-sheng** [comment]
This doc you should update about how to use this dynamic configuration.
https://github.com/apache/skywalking/blob/master/docs/en/setup/backend/dynamic-config.md

And since you are adding new module, https://github.com/apache/skywalking/blob/master/docs/en/setup/backend/configuration-vocabulary.md has to be updated.

Also, it is better to add a whole new menu in the `advanced setup` section about the dynamic logging setup and usage.
image

**Comment by wu-sheng** [comment]
Before doing all the above doc side changes, I want to ask, why this dynamic log should be a new module rather than a feature inside the core module? It looks like no other modules depend on this one, and there is no chance to switch to a new implementation.

**Comment by hanahmily** [comment]
> Before doing all the above doc side changes, I want to ask, why this dynamic log should be a new module rather than a feature inside the core module? It looks like no other modules depend on this one, and there is no chance to switch to a new implementation.

Dynamic logging depends on the `configuration` module. Can a core component support such a constraint?

After this PR, we need to support dynamic logging without `configuration`, a possible path is to inject log4j configuration through RESTful endpoint/GraphQL mutation, that's a new provider more than this one.

**Comment by wu-sheng** [comment]
Why do you need `configuration`? 

> > Before doing all the above doc side changes, I want to ask, why this dynamic log should be a new module rather than a feature inside the core module? It looks like no other modules depend on this one, and there is no chance to switch to a new implementation.
> 
> Dynamic logging depends on the `configuration` module. Can a core component support such a constraint?
> 
> After this PR, we need to support dynamic logging without `configuration`, a possible path is to inject log4j configuration through RESTful endpoint/GraphQL mutation, that's a new provider more than this one.

Core module level HTTP Rest host:port should be a good way to host this service. The existing Query GraphQL isn't. 
Even you want to GraphQL level, it should not be a query protocol. We need to be careful about this kind of feature. CVE and security alert could be easy injected through this way.

File: oap-server/server-bootstrap/src/main/resources/application.yml [hunk 2]
```diff
@@ -510,3 +510,7 @@ configuration-discovery:
 receiver-event:
   selector: ${SW_RECEIVER_EVENT:default}
   default:
+
+logging:
+  selector: ${SW_LOGGING:-}
+  default:
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/application.yml:516]: Legacy change.

**Review by wu-sheng — COMMENTED** [review]
Changelog and doc update please.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/logging/LoggingConfigWatcher.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ * LoggingConfigWatcher watches the change of logging configuration.
+ * Once got the change content, it would apply them to the current logger context.
+ */
+@Slf4j
+public class LoggingConfigWatcher extends ConfigChangeWatcher {
+    private final LoggerContext ctx;
+    private final OapConfiguration originConfiguration;
+    private String content;
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/logging/LoggingConfigWatcher.java:41]:
```suggestion
    private volatile String content;
```
This is being accessed across threads.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/logging/LoggingConfigWatcher.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    public void notify(final ConfigChangeEvent value) {
+        if (EventType.DELETE.equals(value.getEventType())) {
+            this.content = "";
+        } else {
+            this.content = value.getNewValue();
+        }
+        try {
+            boolean applied = updateConfig();
+            if (log.isDebugEnabled() && applied) {
+                log.debug("applied {} B data to logging configuration", this.content.length());
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/logging/LoggingConfigWatcher.java:59]: What does `B` mean?

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/logging/LoggingConfigWatcher.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    @Override
+    public String value() {
+        return this.content;
+    }
+
+    private boolean reconfigure(final String newValue) {
+        if (Strings.isNullOrEmpty(newValue)) {
+            if (ctx.getConfiguration().equals(originConfiguration)) {
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/logging/LoggingConfigWatcher.java:78]:
```suggestion
            if (!ctx.getConfiguration().equals(originConfiguration)) {
```
Not equal should mean reset?

File: oap-server/server-configuration/configuration-k8s-configmap/src/main/java/org/apache/skywalking/oap/server/configuration/configmap/ConfigmapConfigurationWatcherRegister.java [hunk 6]
```diff
@@ -46,9 +45,7 @@ public ConfigmapConfigurationWatcherRegister(ConfigmapConfigurationSettings sett
             if (log.isDebugEnabled()) {
                 log.debug("read config: name:{} ,value:{}", name, value);
             }
-            if (Objects.nonNull(value)) {
-                configTable.add(new ConfigTable.ConfigItem(name, value));
-            }
+            configTable.add(new ConfigTable.ConfigItem(name, value));
```
Comment (wu-sheng) [inline oap-server/server-configuration/configuration-k8s-configmap/src/main/java/org/apache/skywalking/oap/server/configuration/configmap/ConfigmapConfigurationWatcherRegister.java:48]: This should be documented as a bug fix in the changelog.

Comment (hanahmily) [reply oap-server/server-configuration/configuration-k8s-configmap/src/main/java/org/apache/skywalking/oap/server/configuration/configmap/ConfigmapConfigurationWatcherRegister.java:48]: noted

File: docs/en/setup/backend/dynamic-config.md [hunk 7]
```diff
@@ -11,6 +11,7 @@ Right now, SkyWalking supports following dynamic configurations.
 |alarm.default.alarm-settings| The alarm settings, will override `alarm-settings.yml`. | same as [`alarm-settings.yml`](backend-alarm.md) |
 |core.default.apdexThreshold| The apdex threshold settings, will override `service-apdex-threshold.yml`. | same as [`service-apdex-threshold.yml`](apdex-threshold.md) |
 |core.default.endpoint-name-grouping| The endpoint name grouping setting, will override `endpoint-name-grouping.yml`. | same as [`endpoint-name-grouping.yml`](endpoint-grouping-rules.md) |
+|core.default.log4j-xml| The log4j xml configuration, will override `log4j2.xml`. | same as [`log4j2.xml`](https://logging.apache.org/log4j/2.x/manual/configuration.html#XML) |
```
Comment (wu-sheng) [inline docs/en/setup/backend/dynamic-config.md:14]: Please link to our internal doc.

**Comment by wu-sheng** [comment]
Please recheck jdk11 whether it is real issue, or network. I can see this task fails repeatedly.

File: docs/en/setup/backend/dynamic-config.md (hunk 7 above)
Comment (hanahmily) [reply docs/en/setup/backend/dynamic-config.md]: done

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**